proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,710 @@
+  ))
+  (module $child
+    (type $WasiFile (alias $Parent.$WasiFile))
+    (import "wasi_file" (instance (type $WasiFile)))
+  )
+)
+```
+In the binary format, a type alias could be represented by a [De Bruijn index]
+relative to the sequence of parent modules and their respective module/type
+index spaces.
+
+Other than type/module sharing, which can easily be undone by duplication, a
+module A that imports module B is semantically equivalent to module A nesting
+module B. This means tools can merge multiple `.wasm` files into one `.wasm` by
+a trivial, language-independent transformation. Thus, module imports and nested
+modules can serve as a language-independent basis for first expressing a module
+dependency graph and then bundling/linking that dependency graph into a single
+`.wasm` module for portable distribution.
+
+Additionally, on the Web and other hosts that provide runtime instantiation
+APIs, nested modules allow an outer module to bundle module code that is to be
+instantiated dynamically by passing a first-class module reference (created via
+`ref.module`) to the runtime instantiation API.
+
+
+### Module and Instance Exports
+
+Lastly, symmetric to all other kinds of definitions, modules and instances can
+be exported. For example:
+```wasm
+(module
+  (import "a" (module $a ...))
+  (module $b ...)
+  (import "c" (instance $c ...))
+  (instance $d ...)
+
+  (export "e1" (module $a))
+  (export "e2" (module $b))
+  (export "e3" (instance $c))
+  (export "e4" (instance $d))
+)
+```
+Therefore, module and instance types can appear in both the imports and exports
+of module types and instance types.
+
+
+### Summary
+
+To summarize the proposed changes:
+* The `module` field of [`import`] become optional (allowing single-level
+  imports). (How to encode this in the [import section] is an interesting
+  question.)
+* New `module` and `instance` type constructors are added that may be used to
+  define types in the [type section].
+* A new `alias` case is added to type definitions allowing the import of parent
+  modules' type definitions.
+* New `module` and `instance` cases are added to [`importdesc`], with the
+  type indicated by `typeidx`, like `func`.
+* A new `module` section is added which contains a sequence of either module
+  definitions or `alias` definitions (of parents' modules). In terms of the
+  binary format, a nested module definition has the same binary format as a
+  non-nested `.wasm` module. Thus, the [`module`] binary format production
+  becomes recursive.
+* A new `instance` section is added which contains a sequence of instance
+  definitions. Like [`global`] definitions, an `instance` definition contains
+  an `instance` type and an `init` expression which is validated to match.
+* A new `instance.instantiate` instruction is added which is only allowed in
+  `instance` `init` expressions.
+* New `ref.instance` and `ref.module` instructions are added to the general
+  instruction set.
```

Comment:
I wonder if it might be nice to add a few more instructions as well? Some other instructions I could imagine are:

* `instance.get_export "name" (type $ty)` - gets an export named `"name"` with the type `$ty` and returns `(ref $ty)` or traps (or null?). At the top of the explainer you mentioned that it's mirrored decomposition to import a bunch of items or to import one item as an instance, but you can't actually import a module which has a bunch of individual imports, given an instance, without the ability to extract the individual exports.

* `ref.global`, `ref.table`, `ref.memory` - although we have `ref.func`, these might also be necessary to fully satisfy imports for modules? This would be somewhat messy though in that it implies a lot of new types and such.

Reply:
I would also appreciate more detail about how memory imports and exports work in this model. It seems like we would need something like the `ref.memory` as well as the other new reference types Alex mentioned, but in the examples it seems this is somehow being implicitly mangaged.

Reply:
Ah, the key thing here you may have missed is that any imported or defined instances' exports are automatically injected in the containing index spaces.  That's how you can `call $libc.$malloc`.  (This is first mentioned in the "Instance Imports" section.)  With this rule, there should never be a need for a first-class `instance.get_export` because what you want is statically available in an index space.  And, specifically for memory, the exported `libc` memory is getting injected into the containing module's memory index space at index 0, making it the default.

Reply:
Oh right sorry I forgot about that! So yes it looks like `instance.get_export` isn't necessary, but I do think that this may necessitate addition of other instructions like `ref.{global,table,memory}`?

Reply:
Yes, I think you're right: these would enable passing individual globals, tables and memories to `instance.instantiate`.

Reply:
(Added)